spec/cancel_context_spec.rb

```diff
@@ -0,0 +1,69 @@
+# frozen_string_literal: true
+
+require 'iopromise'
+
+RSpec.describe IOPromise::CancelContext do
+  it "cancels IOPromises that are pending after a CancelContext expires" do
+    deferred = nil
+    chained = nil
+
+    IOPromise::CancelContext.with_new_context do
+      deferred = IOPromise::Deferred.new { 123 }
+      chained = deferred.then { 456 }
+
+      expect(deferred).to be_pending
+      expect(deferred).to_not be_cancelled
+      expect(chained).to be_pending
+      expect(chained).to_not be_cancelled
+    end
+
+    expect(deferred).to be_pending
+    expect(deferred).to be_cancelled
+    expect(chained).to be_pending
+    expect(chained).to_not be_cancelled
```

Comment:
I don't have a ton of context, but when reading through this test case I thought that `chained` would have been canceled since its parent, `deferred`, was canceled. I'm curious if that's an intentional design decision or a side effect of the promise implementation?

Reply:
That's correct, we don't propagate cancellation down the chain, and the promises generated with `.then {}` are never registered as IOPromises under the current execution/main loop, so never have the potential to be triggered outside of the request.

Of course, some other code can still trigger these to resolve since they aren't cancelled, but that goes for any `Promise`, and it would be quite a lot of extra housekeeping to register all pending promises in the chain in order to know about them and cancel them. Because of this, we just prevent the IO-based promises themselves from being acted upon once a request completes. In theory, this should provide the same safety guarantee as normal promises.